Further exploration of conditional visibility

drewnoakes opened the pull request:
This gets a few more scenarios working end-to-end. We can tweak the expression language and property arrangement.

Relates to https://github.com/dotnet/project-system/issues/6627.

---

- Fixes calculation of `TargetMultipleFrameworks` (note target framework dropdown not populating)

![multi-target gif](https://user-images.githubusercontent.com/350947/99043389-4b3c1f00-25e2-11eb-88a6-c990e7eba900.gif)

- Makes "Startup Object" conditionally visible based on "Output Type"

![output-type gif](https://user-images.githubusercontent.com/350947/99043068-e84a8800-25e1-11eb-9905-d776b9ebadd9.gif)

- On the packaging page, group licensing options together and set up conditional visibility between them

![license](https://user-images.githubusercontent.com/350947/99043611-93f3d800-25e2-11eb-8dcd-b170b987de6d.gif)

- Use conditional visibility for code signing options

![sign-assembly](https://user-images.githubusercontent.com/350947/99043750-cef60b80-25e2-11eb-8551-664a7a110f74.gif)

drewnoakes committed:
Fix TargetMultipleFrameworksValueProvider logic

It needs to inspect the value of other properties in order to produce
the correct property value.

drewnoakes committed:
Update conditional visibility expressions

- Fixes "TargetMultipleProjects" typo
- Qualify property page names
- Use quotes for literal string values

drewnoakes committed:
Hide "Startup Object" when output type is class library

drewnoakes committed:
Add conditional visibility for resource properties

drewnoakes committed:
Simplify TargetMultipleFrameworksValueProvider

drewnoakes committed:
Add "General" category to package property page

drewnoakes committed:
Group and tidy licensing on packaging page

drewnoakes committed:
Add conditional visibility for signing page

drewnoakes committed:
Remove redundant description text

This is now covered by the conditional visibility behaviour.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/PackagePropertyPage/PackageLicenseKindValueProvider.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+            if (!string.IsNullOrEmpty(await getValue("PackageLicenseExpression")))
+            {
+                return "License";
```
This should be "Expression", not "License".

drewnoakes committed:
Fix license expression value

tmeschter committed:
Use TargetFramework instead of TargetFrameworkMoniker

This fixes an issue where the "Target framework" drop down didn't show the current value.

The drop down is getting the right set of _supported_ values, and if you expand the drop down you will see the display names ".NET Core 2.0", ".NET Core 2.1", ".NET Core 3.0", etc. These display names and the corresponding values (.netcoreapp2.0, .netcoreapp2.1, .netcoreapp3.0, etc.) are ultimately coming from the `SupportedTargetFrameworksProvider`.

However, the _current_ value is coming from the "TargetFrameworkMoniker" property, and that value has the form ".NETCoreApp,Version=3.0" or similar. So the current value doesn't match any of the admissible values.

The solution here is to bind directly to the "TargetFramework" property instead of "TargetFrameworkMoniker". The older property pages use "TargetFrameworkMoniker" due to the requirements of the CSProj project system, and it was carried over into the new pages.

tmeschter commented:
> Fixes calculation of TargetMultipleFrameworks (note target framework dropdown not populating)

I figured this one out. The drop down is getting the right set of _supported_ values, and if you expand the drop down you will see the display names ".NET Core 2.0", ".NET Core 2.1", ".NET Core 3.0", etc. These display names and the corresponding values (.netcoreapp2.0, .netcoreapp2.1, .netcoreapp3.0, etc.) are ultimately coming from the `SupportedTargetFrameworksProvider`.

However, the _current_ value is coming from the "TargetFrameworkMoniker" property, and that value has the form ".NETCoreApp,Version=3.0" or similar. So the current value doesn't match any of the admissible values.

I think the solution here is to bind directly to the "TargetFramework" property instead of "TargetFrameworkMoniker", and I've gone ahead and made that change.

drewnoakes committed:
Treat PackageLicenseFile as a file

Setting the subtype to "file" currently means a "browse" button is shown
in the UI. We may add further treatment for file paths in future.

drewnoakes committed:
Simplify build property page name

drewnoakes committed:
Add check box to control generation of API docs

Relates to dotnet/project-system#368

drewnoakes committed:
Update project on toggle "Target multiple frameworks"

When the synthetic property TargetMultipleFrameworks is updated, move
the value between TargetFramework and TargetFrameworks as appropriate.

drewnoakes committed:
Changing GenerateDocumentationFile cleans up other properties

drewnoakes committed:
Changing PackageLicenseKind cleans up other properties

drewnoakes committed:
Changing ResourceSpecificationKind cleans up other properties

drewnoakes committed:
Improve HelpUrl for GenerateDocumentationFile property

drewnoakes committed:
Add TODO suggesting removing DocumentationFile property

drewnoakes committed:
Specify DependsOn metadata for some properties

drewnoakes committed:
Add Deterministic property

drewnoakes committed:
Replace OutputPath with BaseOutputPath

Allowing control over the per-configuration output path is error prone, as we would allow users to specify the same value for multiple configurations. Using BaseOutputPath avoids that, and configuration-specific subfolders are created according to standard rules.

drewnoakes committed:
Add BaseIntermediateOutputPath property

drewnoakes committed:
Add CheckForOverflowUnderflow property

drewnoakes committed:
Add ErrorReport property

drewnoakes committed:
Add FileAlignment property

drewnoakes committed:
Add NoOp interception base class

Will be used by LinkAction editor properties (hyperlinks and buttons).

drewnoakes committed:
Add hyperlink to "Install other frameworks"

drewnoakes committed:
Add hyperlink "Why can't I select the C# language version?"

drewnoakes committed:
Add hyperlink "Read about SPDX license expressions"

drewnoakes commented:
Added hyperlinks.

9b6d0008099415cc0d06bd926fc2a449190e5046

![image](https://user-images.githubusercontent.com/350947/99232635-4677a500-2846-11eb-93b4-ef867db9a665.png)

b7ed4475727f296c83d8be099e9841862a991d4a

![image](https://user-images.githubusercontent.com/350947/99232680-55f6ee00-2846-11eb-8531-6589d47c537a.png)

636144c6cf5cd438f25a24295799ffa2b93291c0

![image](https://user-images.githubusercontent.com/350947/99232711-6313dd00-2846-11eb-879c-4fc7debff592.png)
